R/utilities-scenarios.R

```diff
@@ -2,311 +2,149 @@
-    } else {
-      warning(paste0(
-        "No individual specific model parameters for individual id '",
-        scenarioConfiguration$individualId, "' found."
-      ))
-    }
-  }
-
-  # Apply test parameters, if specified
-  if (scenarioConfiguration$setTestParameters) {
-    warning("INFO: 'scenarioConfiguration$setTestParameters' is set to TRUE,
-            parameter values defined in 'InputCode/TestParameters.R' will be applied!")
-    # 2DO - remove dependency from getTestParameters() that must be defined outside of the package!
-    params <- extendParameterStructure(
-      parameters = params,
-      newParameters = getTestParameters()
-    )
-  }
-  if (!is.null(customParams)) {
-    params <- extendParameterStructure(
-      parameters = params,
-      newParameters = customParams
-    )
-  }
-
-  # Load simulation
-  simulation <- ospsuite::loadSimulation(filePath = file.path(
-    scenarioConfiguration$projectConfiguration$modelFolder,
-    scenarioConfiguration$modelFile
-  ), loadFromCache = FALSE)
-  # Set the outputs, if new were specified
-  if (!is.null(scenarioConfiguration$outputPaths)) {
-    clearOutputs(simulation)
-    addOutputs(quantitiesOrPaths = scenarioConfiguration$outputPaths, simulation = simulation)
-  }
-  # Set simulation time if defined by the user.
-  if (!is.null(scenarioConfiguration$simulationTime)) {
-    setOutputInterval(
-      simulation = simulation,
-      startTime = 0,
-      endTime = scenarioConfiguration$simulationTime,
-      resolution = scenarioConfiguration$pointsPerMinute
-    )
-  }
-
-  initializeSimulation(
-    simulation = simulation,
-    individualCharacteristics = individualCharacteristics,
-    additionalParams = params,
-    simulateSteadyState = scenarioConfiguration$simulateSteadyState,
-    steadyStateTime = scenarioConfiguration$steadyStateTime,
-    simulationRunOptions = scenarioConfiguration$simulationRunOptions
+createScenarios <- function(scenarioConfigurations, customParams = NULL) {
+  .validateScenarioConfigurations(scenarioConfigurations)
+  .validateParametersStructure(
+    parameterStructure = customParams,
+    argumentName = "customParams",
+    nullAllowed = TRUE
   )
 
-  # Set administration protocols
-  setApplications(simulation = simulation, scenarioConfiguration = scenarioConfiguration)
+  scenarios <- vector("list", length(scenarioConfigurations))
 
-  # Call the custom function.
-  if (!is.null(scenarioConfiguration$customFunction)) {
-    # Set the environment of the custom function to the current environment so it
-    # has access to the `simulation` object
-    environment(scenarioConfiguration$customFunction) <- environment()
-    # Call the custom function with the arguments as defined in the scenario
-    # configuration
-    do.call(what = scenarioConfiguration$customFunction, args = scenarioConfiguration$customFunctionArgs)
+  # For each scenario configuration, create a `Scenario` object
+  for (i in seq_along(scenarioConfigurations)) {
```

Comment:
use `{purrr}` for efficiency.

Something like this should work.

```suggestion
  scenarios <- purrr::map(scenarioConfigurations, ~Scenario$new(.x, customParams = customParams) %>% 
  purrr::set_names(purrr::map(scenarioConfigurations, ~.x$scenarioName)
```